components/patient-measurements/rest/src/main/java/org/phenotips/data/rest/internal/DefaultComputedMeasurementResourceImpl.java

```diff
@@ -57,37 +59,54 @@ public Response getComputedMeasurement(UriInfo uriInfo)
+            String[] dependencies = {params.getFirst("weight"), params.getFirst("height")};
+            value = handleComputations(BMI, dependencies);
+        } else if (USLS.equals(measurement)) {
+            String[] dependencies = {params.getFirst("upperSeg"), params.getFirst("lowerSeg")};
+            value = handleComputations(USLS, dependencies);
         } else {
             throw new WebApplicationException(generateErrorResponse(Response.Status.NOT_FOUND,
                 "Specified measurement type not found."));
         }
+
+        JSONObject resp = new JSONObject();
+        resp.accumulate("value", value);
+
+        return Response.ok(resp, MediaType.APPLICATION_JSON_TYPE).build();
     }
 
     /**
-     * Handle BMI computation.
+     * Handle computations for measurements such as BMI and US:LS.
      *
-     * @param params query parameters
+     * @param measurement the measurement to be computed
+     * @param dependencies dependencies for the computed measurement (eg. weight, height, segment length)
      * @return computed value
      */
-    private double handleBmi(MultivaluedMap<String, String> params)
+    private double handleComputations(String measurement, String[] dependencies) throws IllegalArgumentException
     {
-        String weight = params.getFirst("weight");
-        String height = params.getFirst("height");
-        if (weight == null || height == null) {
-            throw new WebApplicationException(generateErrorResponse(Response.Status.BAD_REQUEST,
-                "Computation arguments were not all provided."));
+        for (String dep : dependencies) {
+            if (dep == null) {
+                throw new WebApplicationException(generateErrorResponse(Response.Status.BAD_REQUEST,
+                    "Computation arguments were not all provided."));
+            }
         }
 
         try {
-            double value = ((BMIMeasurementHandler) this.handlers.get(BMI)).computeBMI(Double.parseDouble(weight),
-                Double.parseDouble(height));
+            double value;
+            if (BMI.equals(measurement)) {
+                // BMI is the measurement to be computed. Pass deps 0 and 1 as the respective height and weight
+                value = ((BMIMeasurementHandler) this.handlers.get(BMI))
+                .computeBMI(Double.parseDouble(dependencies[0]), Double.parseDouble(dependencies[1]));
+            } else if (USLS.equals(measurement)) {
+                // US/LS is the measurement to be computed. Pass deps 0 and 1 as the respective upperSeg and lowerSeg
+                value = ((USLSMeasurementHandler) this.handlers.get(USLS))
+                .computeUSLS(Double.parseDouble(dependencies[0]), Double.parseDouble(dependencies[1]));
+            } else {
+                // The measurement param passed to this method does not match any known computed measurement types
+                throw new IllegalArgumentException();
+            }
 
```

Comment:
This implementation was wrong from the beginning from an architectural pov and did not provide a good example.
There should be no code that is specific to a certain measurement handler, because this approach is not scalable and configurable enough. The goal is to drop in a new handler and have the service work without additional changes. A few suggestions to achieve that:
* The method for getting the value for computed measurements should be generically named `compute` (what else is the BMI handler going to compute if not the BMI?).
* it should get a collection of named values as parameter (that part was good in the original and shouldn't have been changed), and handle the processing internally. Keep in mind that different measurements might have different numbers of dependencies.
* there should be no "if-else" for calling the appropriate handler and no casting to the specific handler's class. It should be as simple as `this.handlers.get(measurement).compute(params)` (inside an `if` or a `try/catch` in case someone tries to call this for a non-computed measurement or a non-existing measurement).

To handle the computed measurements more elegantly I suggest the following:
* the computed measurements should implement a second interface e.g. `ComputedMeasurementHandler` which provides the `compute` method
* the `isComputed` method in the abstract handler should return true if the the measurement implements this interface (which means removing the `isComputed` override from all computed measurements, since we already mark them as computed when we make them implement `ComputedMeasurementHandler`),